File: 0101-0200/0135.Candy.js
//3ms



/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function(ratings) {
    const n = ratings.length;
    let totalCandies = n;
    let i = 1;

    while (i < n) {
        if (ratings[i] === ratings[i - 1]) {
            i++;
            continue;
        }

        let currentPeak = 0;
        while (i < n && ratings[i] > ratings[i - 1]) {
            currentPeak++;
            totalCandies += currentPeak;
            i++;
        }

        if (i === n) {
            return totalCandies;
        }

        let currentValley = 0;
        while (i < n && ratings[i] < ratings[i - 1]) {
            currentValley++;
            totalCandies += currentValley;
            i++;
        }

        totalCandies -= Math.min(currentPeak, currentValley);
    }

    return totalCandies;    
};



//2ms



/**
 * @param {number[]} ratings
 * @return {number}
 */
var candy = function (ratings) {
    let candiesGivenOut = 1;
    let max = ratings[0];
    let maxIdx = 0;
    let count = 1;
    let leeway = 0;

    for (let i = 1; i < ratings.length; i++) {
        if (ratings[i] >= ratings[i - 1]) {
            if (ratings[i] === ratings[i - 1]) {
                candiesGivenOut = 1;
            } else {
                candiesGivenOut++;
            }
            count += candiesGivenOut;
            max = ratings[i];
            maxIdx = i;
            leeway = 0;
        } else {
            if (candiesGivenOut > 1) {
                leeway = candiesGivenOut - 2;
                count += 1;
                candiesGivenOut = 1;
            } else {
                count += i - maxIdx + 1 - (leeway > 0 ? 1 : 0);
                if (leeway > 0) leeway--;
            }
        }
    }

    return count;
};

File: 0101-0200/0135.Candy.java
//3ms


class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;
        int totalCandies = n;
        int i = 1;

        while (i < n) {
            if (ratings[i] == ratings[i - 1]) {
                i++;
                continue;
            }

            int currentPeak = 0;
            while (i < n && ratings[i] > ratings[i - 1]) {
                currentPeak++;
                totalCandies += currentPeak;
                i++;
            }

            if (i == n) {
                return totalCandies;
            }

            int currentValley = 0;
            while (i < n && ratings[i] < ratings[i - 1]) {
                currentValley++;
                totalCandies += currentValley;
                i++;
            }

            totalCandies -= Math.min(currentPeak, currentValley);
        }

        return totalCandies;        
    }
}




//0ms




class Solution {
     static {
        for (int i = 0; i < 140; ++i)
            candy(new int[] { 1, 3, 2 });
            System.gc();
    }
    public static int candy(int[] ratings) {
        int n=ratings.length;
        int i=1;
        int sum=1;
        while(i<n)
        {
            if(ratings[i]==ratings[i-1])
            {
                sum+=1;
                i++;
            }
            int peak=1;
            while(i<n && ratings[i]>ratings[i-1])
            {
                peak++;
                sum+=peak;
                i++;
            }
            int down=1;
            while(i<n && ratings[i]<ratings[i-1])
            {
                sum+=down;
                down++;
                i++;
            }
            if(down>peak)
            {
                sum+=down-peak;
            }
        }
        return sum;
    }
}
